rules/sql-server-memory-pressure/rule.md

```diff
@@ -0,0 +1,94 @@
+---
+type: rule
+archivedreason: 
+title: What to do about SQL Server Memory Pressure?
+guid: f7e8a48d-375e-4c23-9998-9cd9b8c7980c
+uri: sql-performance-memory-pressure
+created: 2023-12-27T07:26:54.0000000Z
+authors:
+  - title: Bryden Oliver
+    url: https://ssw.com.au/people/bryden-oliver
+related: 
+  - identify-sql-performance-sql-server
+  - identify-sql-performance-azure
+  - sql-server-memory-pressure
+  - sql-server-io-pressure
+  - sql-server-cpu-pressure
+---
+So you've identified that your SQL Server is under memory pressure. What can you do about it?
+
+<!--endintro-->
+
+If SQL Server is the primary consumer, then read up on reducing SQL Server memory usage, or whether more memory is appropriate for your workload.
+
+## Identify non SQL engine related usage
+
+Run
+
+```sql
+select * from sys.dm_os_memory_clerks where type='MEMORYCLERK_HOST'
+```
+
+This isolates a few of the SQL processes that aren't part of the SQL Server engine.
+Look for high memory usage for OLE DB providers (MSOLEDBSQL), SQL Native Client (SQLNCLI*) and so on.
+This may indicate using some non core features and you should evaluate whether these are necessary. Non core features are things like running .Net CLR code, translating queries to things like OLE DB and other things that aren't strictly database operations.
+
+## Identify SQL engine related usage
+
+Try running the following query. It categorises the various memory allocations SQL Server has made.
+
+```sql
+SELECT pages_kb, type, name, virtual_memory_committed_kb, awe_allocated_kb
+FROM sys.dm_os_memory_clerks
+ORDER BY pages_kb DESC
+```
+
+This should allow you to identify what is consuming the most memory.
+
+### MEMORYCLERK_SQLQERESERVATIONS
+
+If the memory clerk MEMORYCLERK_SQLQERESERVATIONS is consuming memory, identify queries that are using huge memory grants and optimize them via indexes, rewrite them (remove ORDER by, for example).
+For more information
```

Comment:
**question**: Are the links meant to be dot-points, given that Markdown will smoosh them together without punctuation otherwise?

Reply:
Fixed.